server/queues/tasks/ImportJSONTask.ts

```diff
@@ -152,21 +153,79 @@ export default class ImportJSONTask extends ImportTask {
-    // and replace them out with attachment redirect urls before continuing.
-    for (const document of output.documents) {
-      for (const attachment of output.attachments) {
-        const encodedPath = encodeURI(
-          `/api/attachments.redirect?id=${attachment.externalId}`
-        );
+    // Check all of the attachments we've created against urls and
+    // replace them with the correct redirect urls before continuing.
+    this.replaceAttachmentURLs(output);
 
-        document.text = document.text.replace(
-          new RegExp(escapeRegExp(encodedPath), "g"),
-          `<<${attachment.id}>>`
-        );
+    return output;
+  }
+
+  private replaceAttachmentURLs(output: StructuredImportData) {
+    const attachmentTypes = ["attachment", "image", "video"];
+    const urlRegex = /\/api\/attachments.redirect\?id=(.+)/;
+
+    const attachmentExternalIdMap = output.attachments.reduce(
+      (obj, attachment) => {
+        if (attachment.externalId) {
+          obj[attachment.externalId] = attachment;
+        }
+        return obj;
+      },
+      {} as Record<string, StructuredImportData["attachments"][number]>
+    );
+
+    const getRedirectPath = (existingPath?: string): string | undefined => {
+      if (!existingPath) {
+        return;
       }
-    }
 
-    return output;
+      const match = existingPath.match(urlRegex);
+      if (!match) {
+        return existingPath;
+      }
+
+      const attachment = attachmentExternalIdMap[match[1]];
+      // maintain the existing behaviour of using existingPath when attachment id is not present.
+      return attachment
+        ? Attachment.getRedirectUrl(attachment.id)
+        : existingPath;
+    };
+
+    const transformAttachmentNode = (node: Node): Node => {
+      const json = node.toJSON() as ProsemirrorData;
+      const attrs = json.attrs ?? {};
+
+      if (node.type.name === "attachment") {
+        // attachment node uses 'href' attribute
+        attrs.href = getRedirectPath(attrs.href as string);
+      } else if (node.type.name === "image" || node.type.name === "video") {
+        // image & video nodes use 'src' attribute
+        attrs.src = getRedirectPath(attrs.src as string);
+      }
+
+      json.attrs = attrs;
+      return Node.fromJSON(schema, json);
+    };
+
+    const transformFragment = (fragment: Fragment): Fragment => {
+      const nodes: Node[] = [];
+
+      fragment.forEach((node) => {
+        nodes.push(
+          attachmentTypes.includes(node.type.name)
+            ? transformAttachmentNode(node)
+            : node.copy(transformFragment(node.content))
+        );
+      });
+
+      return Fragment.fromArray(nodes);
+    };
+
+    for (const document of output.documents) {
+      const node = Node.fromJSON(schema, document.data);
+      const transformedNode = node.copy(transformFragment(node.content));
+      document.data = transformedNode;
+      document.text = serializer.serialize(transformedNode);
```

Comment:
Do we end up serializing the node twice here, once on L81 and then again?

Reply:
Umm.. yes.
I think this `for` loop should be moved inside a `if (output.attachments.length) { /* logic to replace attachment urls */ }`.

Reply:
I was thinking that I could make this change once the initial review is over for this PR.
Please let me know when you've reviewed it.

Reply:
Yep, reviewed – this is the only thing that stood out.